File: 프로그래머스/1/42889. 실패율/실패율.java
import java.util.HashMap; 

class Solution {
    public int[] solution(int N, int[] stages) {
        // 실패율 = 스테이지에 도달했으나 아직 클리어하지 못한 플레이어의 수 / 스테이지에 도달한 플레이어 수
        int[] cleared = new int[N+2]; // 몽땅 클리어한 사람은 N+1 stage
        int[] stays = new int[N+2]; // 클리어 못한 사람
        
        HashMap<Integer, Double> failure = new HashMap<>();
        
        int total = stages.length;
        
        // 각 스테이지별 도달한 플레이어 수
        for(int stage :  stages) {
            for (int s = 1; s <= stage; s++) {
                cleared[s]++;
            }
            
            stays[stage]++;
            
        }
        
        for(int s = 1; s <= N; s++) {
            if(cleared[s] == 0) {
                failure.put(s, 0.0);
            }
            else {
                failure.put(s, (stays[s]*1.0)/cleared[s]);
            }
            

        }
        
        
        
        return failure.entrySet().stream().sorted((o1, o2) -> Double.compare(o2.getValue(), o1.getValue())).mapToInt(HashMap.Entry::getKey).toArray();
    }
}

File: 프로그래머스/1/42889. 실패율/실패율.py
def solution(N, stages):
    stay = [0 for _ in range(N+2)]
    # 스테이지에 도달했으나 아직 클리어하지 못한 플레이어의 수 / 스테이지에 도달한 플레이어 수
    # stages[player]==stage / stages[player]>=stage
    
    for stage in stages:
        stay[stage] += 1 # 각 스테이지에 있는 사람 구하기
        
    failures = []
    total = len(stages)
    
    for stage in range(1, N+1):
        reached = total
        not_cleared = stay[stage]
        rate = (not_cleared / reached) if reached > 0 else 0.0
        failures.append((stage, rate))
        total -= not_cleared
        
        failures.sort(key=lambda x: (-x[1], x[0]))
            
    
    

    return [stage for stage, failure in failures ]